### PR TITLE
Support routing in IoTHub

### PR DESCRIPTION
**Pull request by anusapan**
Add new commands for CRUD of Endpoints and Routes in IoTHub.
Add "Test" commands for route.
Update integration test.
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_params.py [hunk 1]
```diff
@@ -11,10 +11,12 @@
                                                 get_resource_name_completion_list,
                                                 get_enum_type,
                                                 get_three_state_flag)
-from azure.mgmt.iothub.models import IotHubSku
-from azure.mgmt.iothubprovisioningservices.models import (IotDpsSku,
-                                                          AllocationPolicy,
-                                                          AccessRightsDescription)
+from azure.mgmt.iothub.models.iot_hub_client_enums import IotHubSku
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_params.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_params.py [hunk 2]
```diff
@@ -11,10 +11,12 @@
...
+from azure.mgmt.iothubprovisioningservices.models.iot_dps_client_enums import (IotDpsSku,
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_params.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_params.py [hunk 3]
```diff
@@ -11,10 +11,12 @@
...
+                                                                               AllocationPolicy,
+                                                                               AccessRightsDescription)
+from azure.cli.command_modules.iot.shared import (EndpointType,
+                                                  RouteSourceType)
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_params.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 4]
```diff
@@ -26,6 +33,7 @@
 
 from azure.cli.command_modules.iot.mgmt_iot_hub_device.lib.iot_hub_device_client import IotHubDeviceClient
 from azure.cli.command_modules.iot.sas_token_auth import SasTokenAuthentication
+from azure.cli.command_modules.iot.shared import (EndpointType)
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 5]
```diff
@@ -508,6 +516,171 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+    if EndpointType.EventHub.value == endpoint_type.lower():
+        hub.properties.routing.endpoints.event_hubs.append(
+            RoutingEventHubProperties(
+                connection_string=connection_string,
+                name=endpoint_name,
+                subscription_id=endpoint_subscription_id,
+                resource_group=endpoint_resource_group
+            )
+        )
+    if EndpointType.ServiceBusQueue.value == endpoint_type.lower():
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 6]
```diff
@@ -508,6 +516,171 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+        hub.properties.routing.endpoints.service_bus_queues.append(
+            RoutingServiceBusQueueEndpointProperties(
+                connection_string=connection_string,
+                name=endpoint_name,
+                subscription_id=endpoint_subscription_id,
+                resource_group=endpoint_resource_group
+            )
+        )
+    if EndpointType.ServiceBusTopic.value == endpoint_type.lower():
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 7]
```diff
@@ -508,6 +516,171 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+        hub.properties.routing.endpoints.service_bus_topics.append(
+            RoutingServiceBusTopicEndpointProperties(
+                connection_string=connection_string,
+                name=endpoint_name,
+                subscription_id=endpoint_subscription_id,
+                resource_group=endpoint_resource_group
+            )
+        )
+    if EndpointType.AzureStorageContainer.value == endpoint_type.lower():
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 8]
```diff
@@ -508,6 +516,171 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+        if container_name is None:
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 9]
```diff
@@ -508,6 +516,171 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+        )
+    )
+    return client.iot_hub_resource.create_or_update(resource_group_name, hub_name, hub, {'IF-MATCH': hub.etag})
+
+
+def iot_hub_route_list(client, hub_name, source_type=None, resource_group_name=None):
+    resource_group_name = _ensure_resource_group_name(client, resource_group_name, hub_name)
+    hub = iot_hub_get(client, hub_name, resource_group_name)
+    return [route for route in hub.properties.routing.routes if route.source.lower() == (source_type.lower()
+                                                                                         if source_type else route.source.lower())]
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 10]
```diff
@@ -508,6 +516,171 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+
+def iot_hub_route_delete(client, hub_name, route_name=None, source_type=None, resource_group_name=None):
+    resource_group_name = _ensure_resource_group_name(client, resource_group_name, hub_name)
+    hub = iot_hub_get(client, hub_name, resource_group_name)
+    if not route_name and not source_type:
+        hub.properties.routing.routes = []
+    if route_name:
+        hub.properties.routing.routes = [route for route in hub.properties.routing.routes
+                                         if route.name.lower() != (route.name.lower()
+                                                                   if route_name is None else route_name.lower())]
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 11]
```diff
@@ -508,6 +516,171 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+    if source_type:
+        hub.properties.routing.routes = [route for route in hub.properties.routing.routes
+                                         if route.source.lower() != (route.source.lower()
+                                                                     if source_type is None else source_type.lower())]
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 12]
```diff
@@ -508,6 +516,171 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+    return client.iot_hub_resource.create_or_update(resource_group_name, hub_name, hub, {'IF-MATCH': hub.etag})
+
+
+def iot_hub_route_update(client, hub_name, route_name, source_type=None, endpoint_name=None, enabled=None, condition=None,
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 13]
```diff
@@ -593,3 +766,35 @@ def _check_dps_name_availability(iot_dps_resource, dps_name):
...
     right_set = set(right_list)  # remove duplicate
     return ",".join(list(right_set))
+
+
+def _delete_routing_endpoints(endpoint_name, endpoint_type, endpoints):
+    if EndpointType.ServiceBusQueue.value == (EndpointType.ServiceBusQueue.value
+                                              if endpoint_type is None else endpoint_type.lower()):
+        new_service_bus_queues = [service_bus_queue for service_bus_queue in endpoints.service_bus_queues
+                                  if service_bus_queue.name.lower() != (service_bus_queue.name.lower()
+                                                                        if endpoint_name is None else endpoint_name.lower())]
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 14]
```diff
@@ -593,3 +766,35 @@ def _check_dps_name_availability(iot_dps_resource, dps_name):
...
+        endpoints.service_bus_queues = new_service_bus_queues
+
+    if EndpointType.ServiceBusTopic.value == (EndpointType.ServiceBusTopic.value
+                                              if endpoint_type is None else endpoint_type.lower()):
+        new_service_bus_topics = [service_bus_topic for service_bus_topic in endpoints.service_bus_topics
+                                  if service_bus_topic.name.lower() != (service_bus_topic.name.lower()
+                                                                        if endpoint_name is None else endpoint_name.lower())]
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 15]
```diff
@@ -593,3 +766,35 @@ def _check_dps_name_availability(iot_dps_resource, dps_name):
...
+        endpoints.service_bus_topics = new_service_bus_topics
+
+    if EndpointType.AzureStorageContainer.value == (EndpointType.AzureStorageContainer.value
+                                                    if endpoint_type is None else endpoint_type.lower()):
+        new_storage_containers = [storage_container for storage_container in endpoints.storage_containers
+                                  if storage_container.name.lower() != (storage_container.name.lower()
+                                                                        if endpoint_name is None else endpoint_name.lower())]
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

File: src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py [hunk 16]
```diff
@@ -593,3 +766,35 @@ def _check_dps_name_availability(iot_dps_resource, dps_name):
...
+        endpoints.storage_containers = new_storage_containers
+
+    if EndpointType.EventHub.value == (EndpointType.EventHub.value
+                                       if endpoint_type is None else endpoint_type.lower()):
+        new_event_hubs = [event_hub for event_hub in endpoints.event_hubs
+                          if event_hub.name.lower() != (event_hub.name.lower()
+                                                        if endpoint_name is None else endpoint_name.lower())]
```
[the remark this answers is not in the document]
Comment (anusapan) [reply src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/custom.py]: Done

**Comment by troydai** [comment]
Thanks for addressing the comments. Please keep debugging the test and make sure CI passes before we can take another look and move towards accepting this PR.

**Comment by anusapan** [comment]
@troydai finally CI passed, Could you please have a another look on this?

**Comment by troydai** [comment]
sure. I'm looking at it.

**Pull request merged**